### DISCUSSION TITLE
How to create a gd.StringName?

### DISCUSSION THREAD

**Discussion by ardawan** [quaadgras/graphics.gd#8] [Q&A] [answered]
i cannot pass the string to the AddResource method. even I tried to cast it to gd.StringName but still it was unsuccessful.

```go
func (w *World) Ready() {             
    w.Class.Super().SetEnvironment(Environment.Instance{})             

    env.AddResource("world1", "res://world1.tres")                      
}
```
error:
```go
env.AddResource undefined (type "grow.graphics/gd/internal/classdb".Environment has no field or method AddResource) 
```

even I've tried to make StringName var but there is no way to add string to it
```go
    var name gd.StringName
    var path gd.Resource 
```

**Comment by Splizard** [comment, marked as answer]
It doesn't look like the Godot Environment class has an `add_resource` method?
https://docs.godotengine.org/en/4.2/classes/class_environment.html

I regards to `StringName`, this is a Godot defined reference type, it can be created with the `StringName` package. The value will be valid until it has been un-used for a frame.

```go
func (w *World) Ready() {
    var name = StringName.New("Go string that will become a StringName")

    // do something with name
}
```

**Note** that for any godot reference types (including string names & classes), the zero value is not valid, ie. it is essentially `nil`. So passing `Environment.Instance{}` does not create a new Environment as you may be expecting.  Use `Environment.New` for this.

**Reply by ardawan** [reply]
Thanks for the explanation.
**What I'm trying to do is to add `new_environment.tres` to the class and see it on the editor.**

as you suggested, I've created this struct and trying to use the reference doc to find out how can I add my environment resource at `res://new-environment.tres` to the placeholder. 
But there is no way for me to find anything that can make this happen.

It would be good if we have some kind of snippet/example for it.

```golang
type World struct {                                       
    classdb.Extension[World, WorldEnvironment.Instance] `gd:"World"`     
    Environment                          WorldEnvronment.Instance
}

func (w *World) Ready(ctx gd.Context) {
    w.Environment.SetEnvironment(??)
}
```
![Screenshot from 2024-02-07 10-14-49](https://github.com/grow-graphics/gd/assets/6112301/4c8824d9-10a1-400a-a437-65216051a4ad)

**Reply by Splizard** [reply]
It's certainly easier to manage resources in this way from the Godot Editor, unless you need to change the environment dynamically.

In such a case you could load the .tres file using the `ResourceLoader` and cast it into the correct type using `gd.As` so it can be passed to the function.
